src/morphosamplers/sampler.py

```diff
@@ -0,0 +1,218 @@
+
+    Parameters
+    ----------
+    grid_shape : Tuple[int, int]
+        Shape of the 2D grid.
+    grid_spacing : Tuple[int, int]
+        Spacing between points in the sampling grid.
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points forming the 2D grid.
+    """
+    # create indices for x and y with correct spacing
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing[0]
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing[1]
+    # convert to coordinate form (x, y, z, 3) (at z = 0)
+    grid = np.stack(np.meshgrid(width, height, 0, indexing="ij"), axis=3)
+    # drop degenerate dim
+    return einops.rearrange(grid, "x y 1 d -> x y d")
+
+
+def generate_3D_grid(
+    grid_shape: Tuple[int, int, int] = (10, 10, 10),
+    grid_spacing: Tuple[int, int, int] = (1, 1, 1),
+) -> np.ndarray:
+    """
+    Generate a 3D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, has shape (*grid_shape, 3)
+    and spacing grid_spacing between neighboring points.
+
+    Parameters
+    ----------
+    grid_shape : Tuple[int, int, int]
+        Shape of the 3D grid.
+    grid_spacing : Tuple[int, int, int]
+        Spacing between points in the sampling grid.
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points forming the 2D grid.
+    """
+    # create indices for x and y with correct spacing
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing[0]
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing[1]
+    depth = np.linspace(-1, 1, grid_shape[2]) * grid_shape[2] * grid_spacing[2]
+    # convert to stack form (x, y, z, 3)
+    grid = np.stack(np.meshgrid(width, height, depth, indexing="ij"), axis=3)
+    return grid
+
+
+def generate_sampling_coordinates(
+    sampling_grid: np.ndarray, positions: np.ndarray, orientations: Rotation
+) -> np.ndarray:
+    """
+    Copy and transform a given sampling grid onto a set of positions and orientations.
+
+    Returns an (n, *grid_shape, 3) grid of sampling points.
+
+    Parameters
+    ----------
+    sampling_grid : np.ndarray
+        Grid of points to be copied and transformed.
+    positions : np.ndarray
+        Array of coordinates used to shift sampling grids.
+    orientations : Rotation
+        Rotations to be applied to each grid before shifts.
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points combining all the transformed grids.
+    """
+    rotated = []
+    grid_shape = sampling_grid.shape
+    grid_coords = sampling_grid.reshape(-1, 3)
+    # apply each orientation to the grid and store the result
+    for ori in orientations:
+        rotated.append(ori.apply(grid_coords))
+    # shift each rotated
+    rotated_shifted = np.stack(rotated, axis=1) + positions
+    return rotated_shifted.reshape(-1, *grid_shape)
+
+
+def sample_volume_with_coordinates(
```

Comment:
```suggestion
def sample_volume_at_coordinates(
```
this makes more sense to me